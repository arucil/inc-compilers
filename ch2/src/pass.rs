pub mod assign_home;
pub mod explicate_control;
pub mod instruction_selection;
pub mod partial_evaluation;
pub mod patch_instructions;
pub mod remove_complex_operands;
pub mod uniquify;
pub mod perilogue;