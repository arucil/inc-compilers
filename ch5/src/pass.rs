mod explicate_control;
mod instruction_selection;
pub mod liveness_analysis;
mod remove_complex_operands;
pub mod typecheck;
pub mod uncover_get;
pub mod perilogue;
