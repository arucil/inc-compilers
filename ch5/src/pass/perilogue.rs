use asm::{Arg, Block, Instr, Label, Program};
use ch4::pass::register_allocation::Info;

pub fn add_perilogue(prog: &mut Program<Info>) {
  add_prologue(prog);
  add_epilogue(prog);
}

fn add_prologue(prog: &mut Program<Info>) {
  use asm::Reg::*;
  use Instr::*;
  let stack_space =
    prog.info.stack_space + prog.info.used_callee_saved_regs.len() * 8;
  let stack_space =
    ((stack_space + 15) & !15) - prog.info.used_callee_saved_regs.len() * 8;
  let mut code = vec![
    Push(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Reg(Rsp),
      dest: Arg::Reg(Rbp),
    },
    Sub {
      src: Arg::Imm(stack_space as i64),
      dest: Arg::Reg(Rsp),
    },
  ];
  for &reg in &prog.info.used_callee_saved_regs {
    code.push(Push(Arg::Reg(reg)));
  }
  code.push(Jmp(Arg::Label(Label::Start)));
  let block = Block { global: true, code };
  prog.blocks.push((Label::EntryPoint, block));
}

fn add_epilogue(prog: &mut Program<Info>) {
  use asm::Reg::*;
  use Instr::*;
  let mut code: Vec<Instr> = prog
    .info
    .used_callee_saved_regs
    .iter()
    .rev()
    .map(|&reg| Pop(Arg::Reg(reg)))
    .collect();
  code.extend_from_slice(&[
    Mov {
      src: Arg::Reg(Rbp),
      dest: Arg::Reg(Rsp),
    },
    Pop(Arg::Reg(Rbp)),
    Mov {
      src: Arg::Imm(60),
      dest: Arg::Reg(Rax),
    },
    Mov {
      src: Arg::Imm(0),
      dest: Arg::Reg(Rdi),
    },
    Syscall,
  ]);
  let block = Block {
    global: false,
    code,
  };
  prog.blocks.push((Label::Conclusion, block));
}
