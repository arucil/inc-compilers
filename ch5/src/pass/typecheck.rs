use ast::{Program, Type};
use ch4::pass::typecheck::TypeChecker;
use support::CompileError;

pub type Result<T> = std::result::Result<T, CompileError>;

pub fn typecheck(prog: Program) -> Result<Program<String, Type>> {
  let mut fun_defs = prog.fun_defs;
  let mut checker = TypeChecker::new(prog.type_defs, &mut fun_defs)?;
  let fun_defs = fun_defs
    .into_iter()
    .map(|(name, fun)| checker.typecheck_fun(name, fun))
    .collect::<Result<_>>()?;
  let range = prog.body.range;
  let body = checker.typecheck(prog.body)?;
  if body.ty != Type::Void {
    return Err(CompileError {
      range,
      message: format!("expected Void, found {:?}", body.ty),
    });
  }
  Ok(Program {
    fun_defs,
    type_defs: Default::default(),
    body,
    types: checker.types,
  })
}
