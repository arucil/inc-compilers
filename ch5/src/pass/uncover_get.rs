use ast::{Exp, ExpKind, IdxVar, Program};
use std::collections::HashSet;

pub fn uncover_get<TYPE>(prog: Program<IdxVar, TYPE>) -> Program<IdxVar, TYPE> {
  Program {
    body: prog.body.into_iter().map(uncover_exp).collect(),
    types: prog.types,
  }
}

fn uncover_exp<TYPE>(exp: Exp<IdxVar, TYPE>) -> Exp<IdxVar, TYPE> {
  let mut set_vars = HashSet::new();
  collect_set_vars(&exp, &mut set_vars);
  mark_mutable_vars(exp, &set_vars)
}

fn mark_mutable_vars<TYPE>(
  exp: Exp<IdxVar, TYPE>,
  set_vars: &HashSet<IdxVar>,
) -> Exp<IdxVar, TYPE> {
  match exp.kind {
    ExpKind::Var(var) => {
      if set_vars.contains(&var) {
        Exp {
          kind: ExpKind::Get(var),
          range: exp.range,
          ty: exp.ty,
        }
      } else {
        Exp {
          kind: ExpKind::Var(var),
          range: exp.range,
          ty: exp.ty,
        }
      }
    }
    ExpKind::Int(_) | ExpKind::Str(_) | ExpKind::Bool(_) | ExpKind::Void => exp,
    ExpKind::Let { var, init, body } => Exp {
      kind: ExpKind::Let {
        var,
        init: box mark_mutable_vars(*init, set_vars),
        body: box mark_mutable_vars(*body, set_vars),
      },
      range: exp.range,
      ty: exp.ty,
    },
    ExpKind::Begin { seq, last } => Exp {
      kind: ExpKind::Begin {
        seq: seq
          .into_iter()
          .map(|exp| mark_mutable_vars(exp, set_vars))
          .collect(),
        last: box mark_mutable_vars(*last, set_vars),
      },
      range: exp.range,
      ty: exp.ty,
    },
    ExpKind::If { cond, conseq, alt } => Exp {
      kind: ExpKind::If {
        cond: box mark_mutable_vars(*cond, set_vars),
        conseq: box mark_mutable_vars(*conseq, set_vars),
        alt: box mark_mutable_vars(*alt, set_vars),
      },
      range: exp.range,
      ty: exp.ty,
    },
    ExpKind::Prim { op, args } => Exp {
      kind: ExpKind::Prim {
        op,
        args: args
          .into_iter()
          .map(|exp| mark_mutable_vars(exp, set_vars))
          .collect(),
      },
      range: exp.range,
      ty: exp.ty,
    },
    ExpKind::Print(args) => Exp {
      kind: ExpKind::Print(
        args
          .into_iter()
          .map(|exp| mark_mutable_vars(exp, set_vars))
          .collect(),
      ),
      range: exp.range,
      ty: exp.ty,
    },
    ExpKind::Set { var, exp: exp1 } => Exp {
      kind: ExpKind::Set {
        var,
        exp: box mark_mutable_vars(*exp1, set_vars),
      },
      range: exp.range,
      ty: exp.ty,
    },
    ExpKind::NewLine => exp,
    ExpKind::While { cond, body } => Exp {
      kind: ExpKind::While {
        cond: box mark_mutable_vars(*cond, set_vars),
        body: box mark_mutable_vars(*body, set_vars),
      },
      range: exp.range,
      ty: exp.ty,
    },
    ExpKind::Get(_) => unreachable!(),
  }
}

fn collect_set_vars<TYPE>(
  exp: &Exp<IdxVar, TYPE>,
  set_vars: &mut HashSet<IdxVar>,
) {
  match &exp.kind {
    ExpKind::Int(_)
    | ExpKind::Str(_)
    | ExpKind::Bool(_)
    | ExpKind::Void
    | ExpKind::Var(_) => {}
    ExpKind::Let { init, body, .. } => {
      collect_set_vars(init, set_vars);
      collect_set_vars(body, set_vars);
    }
    ExpKind::Begin { seq, last } => {
      for exp in seq {
        collect_set_vars(exp, set_vars);
      }
      collect_set_vars(last, set_vars);
    }
    ExpKind::If { cond, conseq, alt } => {
      collect_set_vars(cond, set_vars);
      collect_set_vars(conseq, set_vars);
      collect_set_vars(alt, set_vars);
    }
    ExpKind::Prim { args, .. } => {
      for exp in args {
        collect_set_vars(exp, set_vars);
      }
    }
    ExpKind::Print(args) => {
      for exp in args {
        collect_set_vars(exp, set_vars);
      }
    }
    ExpKind::Set { var, exp } => {
      set_vars.insert(var.1.clone());
      collect_set_vars(exp, set_vars);
    }
    ExpKind::NewLine => {}
    ExpKind::While { cond, body } => {
      collect_set_vars(cond, set_vars);
      collect_set_vars(body, set_vars);
    }
    ExpKind::Get(_) => unreachable!(),
  }
}
