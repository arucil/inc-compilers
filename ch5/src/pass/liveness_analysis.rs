use asm::{Instr, Label, Program};
use ast::IdxVar;
use ch2::pass::select_instruction::Info as OldInfo;
use ch3::location_set::{LocationSet, VarStore};
use ch3::pass::liveness_analysis::{AnalysisState, Info as NewInfo};
use indexmap::IndexMap;
use petgraph::graph::NodeIndex;
use petgraph::{Direction, Graph};
use std::collections::{HashMap, VecDeque};
use std::hash::Hash;

/// `label_live` is a map from labels to sets of live locations before the first
/// instruction of the blocks.
pub fn analyze_liveness(
  prog: Program<OldInfo, IdxVar>,
  label_live: HashMap<Label, LocationSet>,
) -> Program<NewInfo, IdxVar> {
  let mut var_store = VarStore::new();
  for var in &prog.info.locals {
    var_store.insert(var.clone());
  }

  let blocks: HashMap<_, _> = prog
    .blocks
    .iter()
    .map(|(label, block)| (*label, block))
    .collect();
  let mut graph = Graph::<Label, ()>::new();
  let mut nodes = HashMap::<Label, NodeIndex>::new();
  for (label, _) in &prog.blocks {
    let ix = graph.add_node(*label);
    nodes.insert(*label, ix);
  }

  for node in nodes.values() {
    let last = blocks[&graph[*node]].code.last().unwrap().clone();
    if let Instr::Jmp(label) = last {
      if let Some(&node1) = nodes.get(&label) {
        graph.add_edge(node1, *node, ());
      }
      let code = &blocks[&graph[*node]].code;
      if code.len() > 1 {
        if let Instr::JumpIf { label, .. } = code[code.len() - 2].clone() {
          graph.add_edge(nodes[&label], *node, ());
        }
      }
    }
  }

  let state = AnalysisState {
    var_store: &var_store,
  };
  let mut live = IndexMap::<Label, Vec<LocationSet>>::new();

  analyze_dataflow(
    label_live,
    graph,
    |mapping, label, _| {
      let l = state.block_liveness(blocks[&label], mapping);
      let output = l[0].clone();
      live.insert(label, l);
      output
    },
    LocationSet::new(),
    |mut set1, set2| {
      set1 |= set2;
      set1
    },
  );

  Program {
    info: NewInfo {
      locals: prog.info.locals,
      live,
      var_store,
    },
    constants: prog.constants,
    blocks: prog.blocks,
  }
}

pub fn analyze_dataflow<N, T, F, J>(
  mut mapping: HashMap<N, T>,
  g: Graph<N, ()>,
  mut transfer: F,
  bottom: T,
  join: J,
) -> HashMap<N, T>
where
  N: Copy + Hash + Eq,
  T: Clone + PartialEq,
  F: FnMut(&HashMap<N, T>, N, T) -> T,
  J: Fn(T, &T) -> T,
{
  for &label in g.node_weights() {
    mapping.insert(label, bottom.clone());
  }

  let mut worklist: VecDeque<_> = g.node_indices().collect();
  while let Some(node_ix) = worklist.pop_front() {
    let label = g[node_ix];
    let input = g
      .neighbors_directed(node_ix, Direction::Incoming)
      .map(|node_ix| &mapping[&g[node_ix]])
      .fold(bottom.clone(), &join);
    let output = transfer(&mapping, label, input);
    if output != mapping[&label] {
      mapping.insert(label, output);
      for node_ix in g.neighbors_directed(node_ix, Direction::Outgoing) {
        worklist.push_back(node_ix);
      }
    }
  }

  mapping
}

#[cfg(test)]
mod tests {
  use super::*;
  use asm::Label;
  use indexmap::IndexSet;
  use insta::assert_snapshot;
  use maplit::hashmap;

  trait ShowLiveness {
    fn show(&self) -> String;
  }

  impl ShowLiveness for Program<NewInfo, IdxVar> {
    fn show(&self) -> String {
      let mut buf = String::new();
      for (label, block) in &self.blocks {
        buf += &format!("{:?}:\n", label);
        let live = &self.info.live[label];
        for (i, l) in live.iter().enumerate().take(block.code.len()) {
          buf += "                    ";
          l.write(&mut buf, &self.info.var_store).unwrap();
          buf += "\n";
          buf += &format!("    {:?}", block.code[i]);
          buf += "\n";
        }
        buf += "                    ";
        live
          .last()
          .unwrap()
          .write(&mut buf, &self.info.var_store)
          .unwrap();
        buf += "\n\n";
      }
      buf
    }
  }

  #[test]
  fn example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        mov v, 1
        mov w, 42
        mov x, v
        add x, 7
        mov y, x
        mov z, x
        add z, w
        mov t, y
        neg t
        mov rax, z
        add rax, t
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn push_pop() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        push x
        mov w, rbx
        pop rbx
        add x, w
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn call() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        pop rdi
        pop rsi
        push x
        mov w, rbx
        call foo, 3
        add w, rax
        jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn epilogue() {
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        mov rsp, rbp
        pop rbp
        call print_int
        call print_newline
        mov rax, 60
        mov rdi, 0
        syscall
    "#,
    );
    let label_live = HashMap::new();
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn nested_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
      start:
        call read_int
        mov x, rax
        call read_int
        mov y, rax
        cmp x, 1
        jl block2
        jmp block3
      block0:
        mov rax, y
        add rax, 2
        jmp conclusion
      block1:
        mov rax, z
        add rax, 10
        jmp conclusion
      block2:
        cmp x, 0
        je block0
        jmp block1
      block3:
        cmp x, 2
        je block0
        jmp block1
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn if_in_init() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov tmp.0, rax
    cmp tmp.0, 3
    jge block3
    jmp block4
block0:
    mov rax, 2
    jmp conclusion
block1:
    mov rax, 41
    jmp conclusion
block2:
    cmp x.0, 10
    sete al
    movzx tmp.1, al
    cmp tmp.1, 0
    je block1
    jmp block0
block3:
    mov x.0, 10
    jmp block2
block4:
    mov x.0, 77
    jmp block2

    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn complex_if() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    call read_int
    mov x.0, rax
    cmp x.0, 100
    jg block8
    jmp block9
block0:
    mov rax, -1
    jmp conclusion
block1:
    mov rax, y.1
    neg rax
    jmp conclusion
block2:
    mov tmp.1, y.1
    sub tmp.1, x.0
    cmp tmp.1, 10
    jl block0
    jmp block1
block3:
    cmp x.0, y.1
    jl block2
    jmp block1
block4:
    mov tmp.0, x.0
    sub tmp.0, y.1
    cmp tmp.0, 10
    jl block0
    jmp block3
block5:
    mov rax, 5000
    jmp conclusion
block6:
    mov rax, x.0
    jmp conclusion
block7:
    cmp x.0, 60
    jl block5
    jmp block6
block8:
    call read_int
    mov y.1, rax
    cmp x.0, y.1
    jge block4
    jmp block3
block9:
    cmp x.0, 40
    jg block7
    jmp block6
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }

  #[test]
  fn ch5_example_in_book() {
    use asm::Reg::*;
    let blocks = asm::parse_blocks(
      |s| IdxVar::new(s),
      r#"
start:
    mov sum.0, 0
    mov i.0, 5
    jmp block5
block5:
    mov tmp.3, i.0
    cmp tmp.3, 0
    jg block7
    jmp block8
block7:
    add sum.0, i.0
    mov tmp.4, 1
    neg tmp.4
    add tmp.4, i.0
    jmp block5
block8:
    mov rax, 27
    add rax, sum.0
    jmp conclusion
    "#,
    );
    let label_live = hashmap! {
      Label::Conclusion => LocationSet::regs([Rax, Rsp])
    };
    let prog = Program {
      info: OldInfo {
        locals: IndexSet::new(),
      },
      constants: Default::default(),
      blocks,
    };
    let result = analyze_liveness(prog, label_live);

    assert_snapshot!(result.show());
  }
}
