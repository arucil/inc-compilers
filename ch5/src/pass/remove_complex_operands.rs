#[cfg(test)]
mod tests {
  use super::super::*;
  use ast::*;
  use ch2::pass::remove_complex_operands;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use insta::assert_snapshot;

  #[test]
  fn set_get() {
    let prog =
      parse(r#"(let ([x 2] [y 0]) (+ y (+ x (begin (set! x 40) x))))"#)
        .unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = uncover_get::uncover_get(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn complex() {
    let prog = ast::parse(
      r#"
(let ([x 0] [y 10] [step (- y 7)] [z (read)])
  (print "start")
  (print (if (not (eq? y z)) "y!=z " "y=z") (eq? z y))
  (while (< x y)
    (let ([t (+ x step)])
      (set! x t)
      (print
        "?="
        (if (and (> x 5) (< x 9))
          (begin
            (print "neg!")
            (set! z (+ z 1))
            (- step))
          z)))))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let result = remove_complex_operands::remove_complex_operands(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
