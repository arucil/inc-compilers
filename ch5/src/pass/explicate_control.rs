#[cfg(test)]
mod tests {
  use super::super::*;
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::shrink;
  use ch4::pass::uniquify;
  use insta::assert_snapshot;

  #[test]
  fn begin() {
    let prog = ast::parse(
      r#"
(let ([x 1] [y (read)])
  (while (> x 10)
    (set! x (begin (print "x=" x) (+ x (- y 1))))
    (print (> x (- y 1))))
  (begin 1 2 3)
  (let ([k (+ x 1)])
    (print k)
    k)
  (+ 7 (+ (read) (if (begin (set! y 37) (not (eq? y x))) 11 23)))
  (+ x y)
  (void))
      "#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn assign_void() {
    let prog = ast::parse(
      r#"(let ([x (void)] [y 1]) (set! x (print)) (set! x (set! y 10)))"#,
    )
    .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = shrink::shrink(prog);
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let result = explicate_control::explicate_control(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
