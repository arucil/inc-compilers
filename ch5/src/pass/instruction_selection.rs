#[cfg(test)]
mod tests {
  use super::super::*;
  use ast::*;
  use ch2::pass::remove_complex_operands;
  use ch4::pass::explicate_control;
  use ch4::pass::instruction_selection;
  use ch4::pass::uniquify;
  use insta::assert_snapshot;

  #[test]
  fn single_read() {
    let prog = parse(r#"(begin (read) (void))"#).unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }

  #[test]
  fn assign_void() {
    let prog =
      parse(r#"(let ([x (void)] [y (set! x (print 3))]) (set! x (print #t)))"#)
        .unwrap();
    let prog = typecheck::typecheck(prog).unwrap();
    let prog = uniquify::uniquify(prog);
    let prog = remove_complex_operands::remove_complex_operands(prog);
    let prog = explicate_control::explicate_control(prog);
    let result = instruction_selection::select_instruction(prog);
    assert_snapshot!(result.to_string_pretty());
  }
}
