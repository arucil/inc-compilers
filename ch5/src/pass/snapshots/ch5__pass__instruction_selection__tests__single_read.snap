---
source: ch5/src/pass/instruction_selection.rs
expression: result.to_string_pretty()
---
locals: {}

constants: {}

.start:
    call rt_read_int
    jmp .conclusion

